File: Informatică și tehnologia informației/proba practica/rezolvare_2025/Subiectul_I/b.pas
program DivizoriB;

type
  TTablou = array[1..100] of integer;

var
  fin, fout: text;
  n, i, x, count: integer;
  v: TTablou;

// Subprogramul nrdiv primeste ca parametru un numar natural n si returneaza numarul de divizori ai sai.
function nrdiv(n: integer): integer;
var
  i, cnt: integer;
begin
  cnt := 0;
  i := 1;
  while i * i <= n do
  begin
    if n mod i = 0 then
    begin
      cnt := cnt + 1;
      if i * i <> n then
        cnt := cnt + 1;
    end;
    i := i + 1;
  end;
  nrdiv := cnt;
end;

// Subprogramul ordonare primeste numarul de elemente n si tabloul v si il ordoneaza descrescator.
procedure ordonare(n: integer; var v: TTablou);
var
  i, j, temp: integer;
begin
  for i := 1 to n - 1 do
    for j := i + 1 to n do
      if v[i] < v[j] then
      begin
        temp := v[i];
        v[i] := v[j];
        v[j] := temp;
      end;
end;

begin
  assign(fin, 'divizori.in');
  reset(fin);
  assign(fout, 'divizori.out');
  rewrite(fout);
  
  if eof(fin) then
  begin
    close(fin);
    close(fout);
    exit;
  end;
  
  read(fin, n);
  count := 0;
  for i := 1 to n do
  begin
    read(fin, x);
    if nrdiv(x) mod 2 <> 0 then
    begin
      count := count + 1;
      v[count] := x;
    end;
  end;
  
  if count = 0 then
    write(fout, 'nu exista')
  else
  begin
    ordonare(count, v);
    for i := 1 to count do
    begin
      write(fout, v[i]);
      if i < count then
        write(fout, ' ');
    end;
    writeln(fout);
  end;
  
  close(fin);
  close(fout);
end.

File: Informatică și tehnologia informației/proba practica/rezolvare_2025/Subiectul_I/b.cpp
#include <iostream>
#include <fstream>

using namespace std;

// Subprogramul nrdiv primeste ca parametru un numar natural n si returneaza numarul de divizori ai sai.
int nrdiv(int n) {
    int cnt = 0;
    int i = 1;
    for (; i * i <= n; ++i) {
        if (n % i == 0) {
            cnt++;
            if (i * i != n) {
                cnt++;
            }
        }
    }
    return cnt;
}

// Subprogramul ordonare primeste numarul de elemente n si tabloul v si il ordoneaza descrescator.
void ordonare(int n, int v[]) {
    for (int i = 0; i < n - 1; ++i) {
        for (int j = i + 1; j < n; ++j) {
            if (v[i] < v[j]) {
                int aux = v[i];
                v[i] = v[j];
                v[j] = aux;
            }
        }
    }
}

int main() {
    ifstream fin("divizori.in");
    ofstream fout("divizori.out");
    
    if (!fin) {
        cerr << "Eroare la deschiderea fisierului divizori.in" << endl;
        return 1;
    }
    
    int n;
    if (!(fin >> n)) return 0;
    
    int v[100];
    int count = 0;
    
    for (int i = 0; i < n; ++i) {
        int x;
        fin >> x;
        // Daca numarul de divizori este impar, il retinem in tablou
        if (nrdiv(x) % 2 != 0) {
            v[count++] = x;
        }
    }
    
    if (count == 0) {
        fout << "nu exista\n";
    } else {
        ordonare(count, v);
        for (int i = 0; i < count; ++i) {
            fout << v[i] << (i == count - 1 ? "" : " ");
        }
        fout << "\n";
    }
    
    return 0;
}
